File: tools/drive.py
#!/usr/bin/env python3
"""What the iCloud Drive pull last fetched, per library. Status only.

**The pull itself is not a tool and deliberately never will be.** `bin/icloud_drive_fetch.py`
fetches and stages files as this account, on whatever schedule you give it, and whatever consumes
those files runs as its own account with its own access. Separate accounts on purpose: the fetcher
cannot reach what reads the staged files, so a compromised browser cannot rewrite data it was
never meant to touch. An MCP tool that could fetch, move and import would collapse that into one.

So what a client gets is the one thing it needs and cannot infer: whether the pull is current. A
week of data that stops on Tuesday is a pull that has not run, or a library the source app stopped
writing to, and that is a different fact from a week with nothing to record. Reporting freshness
here means whatever reads those files does not have to guess why they are short.
"""
import json
from datetime import datetime, timezone

# Which libraries this install pulls, and where they are staged, both from the environment. The
# etag file is keyed by Apple's opaque `docwsid`, so it can say how many files are tracked in
# total and nothing about which library they belong to; the staging tree is what carries that.
from icloud_lib.drive_libraries import ETAGS, LIBRARIES, STAGING

from app import mcp


def _age(ts: float | None) -> dict:
    if ts is None:
        return {"at": None, "hours_ago": None}
    when = datetime.fromtimestamp(ts, timezone.utc)
    return {"at": when.isoformat(timespec="seconds"),
            "hours_ago": round((datetime.now(timezone.utc) - when).total_seconds() / 3600, 1)}


@mcp.tool()
def drive_status() -> dict:
    """When the iCloud Drive pull last ran, and what it holds per library.

    Read this before concluding that data pulled off Drive is missing rather than stale. Compare
    the age of the last run against how often the pull is scheduled: a gap much larger than that
    interval means it has not run.
    """
    try:
        etags = json.loads(ETAGS.read_text())
    except OSError:
        etags = {}
        last_run = None
    except json.JSONDecodeError as exc:
        return {"error": f"{ETAGS} is not readable as JSON: {exc}",
                "note": "the pull writes this file at the end of every run, so an unparseable "
                        "one means a run was interrupted partway"}
    else:
        last_run = ETAGS.stat().st_mtime

    per_library = {}
    for lib in LIBRARIES:
        staged = STAGING / lib["dest"]
        times = [p.stat().st_mtime for p in staged.rglob("*") if p.is_file()] \
            if staged.is_dir() else []
        per_library[lib["name"]] = {
            "kind": lib["kind"],
            "staged_under": lib["dest"],
            # Zero is the healthy answer. The sync script moves everything out of staging after
            # every fetch, so a non-zero count means a run stopped between the fetch and the
            # import, which is worth saying rather than hiding.
            "staged_now": len(times),
            "newest_staged_file": _age(max(times) if times else None),
        }

    stale = last_run is not None and (
        datetime.now(timezone.utc).timestamp() - last_run) > 25 * 3600
    return {
        "last_pull": _age(last_run),
        # Named rather than implied: the answer to "is this stale" should not require the reader
        # to do the arithmetic, and 25 hours is one daily run plus an hour of slack.
        "stale": stale if last_run is not None else None,
        "schedule": "a daily timer, persistent, so a missed run catches up on boot",
        # Total, not per library: the etag file is keyed by Apple's opaque docwsid.
        "files_tracked": len(etags),
        "libraries": per_library,
        "note": "Status only. The pull runs as this account from a timer and the import runs as "
                "the owning service's account; no tool here can trigger either, because the "
                "fetcher deliberately cannot reach those databases. `last_pull` is the freshness "
                "signal; `staged_now` is 0 in normal operation because the sync empties staging "
                "after every run.",
    }

File: internal/drive/drive.go
// Package drive is what the iCloud Drive pull last fetched, per library.
// Status only.
//
// The pull itself is not a tool and deliberately never will be: the fetcher
// runs as this account on a schedule, and whatever consumes the staged
// files runs as its own account. An MCP tool that could fetch, move, and
// import would collapse that separation into one.
package drive

import (
	"context"
	"encoding/json"
	"math"
	"os"
	"path/filepath"
	"time"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"

	"github.com/sjdonado/icloud-headless-mcp/internal/config"
	"github.com/sjdonado/icloud-headless-mcp/internal/dvlibraries"
	"github.com/sjdonado/icloud-headless-mcp/internal/mcpserver"
)

// age renders a timestamp the way drive_status rows read.
func age(ts *time.Time, now time.Time) map[string]any {
	if ts == nil {
		return map[string]any{"at": nil, "hours_ago": nil}
	}
	return map[string]any{
		"at":        ts.UTC().Format("2006-01-02T15:04:05Z07:00"),
		"hours_ago": math.Round(now.Sub(*ts).Hours()*10) / 10,
	}
}

// Status reports when the pull last ran and what is staged per library.
// It cannot trigger the pull, and neither can anything else here.
func Status(cfg *config.Config, now time.Time) (map[string]any, error) {
	etags := map[string]any{}
	raw, err := os.ReadFile(cfg.DriveEtags)
	var lastRun *time.Time
	if err == nil {
		var parsed map[string]any
		if jerr := json.Unmarshal(raw, &parsed); jerr != nil {
			return map[string]any{
				"error": cfg.DriveEtags + " is not readable as JSON: " + jerr.Error(),
				"note": "the pull writes this file at the end of every run, so an unparseable " +
					"one means a run was interrupted partway",
			}, nil
		}
		etags = parsed
		if info, serr := os.Stat(cfg.DriveEtags); serr == nil {
			t := info.ModTime()
			lastRun = &t
		}
	}
	perLibrary := map[string]any{}
	for _, lib := range dvlibraries.ParseLibraries(cfg.DriveLibs) {
		staged := filepath.Join(cfg.DriveStaging, lib.Dest)
		var newest *time.Time
		count := 0
		walkErr := filepath.Walk(staged, func(p string, info os.FileInfo, err error) error {
			if err != nil {
				return err
			}
			if info.IsDir() {
				return nil
			}
			count++
			mt := info.ModTime()
			if newest == nil || mt.After(*newest) {
				newest = &mt
			}
			return nil
		})
		// A missing staging dir is normal (nothing pulled yet); any other
		// walk failure is a real error, not a healthy zero.
		if walkErr != nil && !os.IsNotExist(walkErr) {
			return nil, walkErr
		}
		perLibrary[lib.Name] = map[string]any{
			"kind":         lib.Kind,
			"staged_under": lib.Dest,
			// Zero is the healthy answer: the sync moves everything out of
			// staging after every fetch, so a non-zero count means a run
			// stopped between the fetch and the import.
			"staged_now":         count,
			"newest_staged_file": age(newest, now),
		}
	}
	var stale any
	if lastRun != nil {
		stale = now.Sub(*lastRun) > 25*time.Hour
	}
	return map[string]any{
		"last_pull": age(lastRun, now),
		// Named rather than implied, so the reader never does the
		// arithmetic; 25 hours is one daily run plus an hour of slack.
		"stale":    stale,
		"schedule": "a daily timer, persistent, so a missed run catches up on boot",
		// Total, not per library: the etag file is keyed by Apple's
		// opaque docwsid.
		"files_tracked": len(etags),
		"libraries":     perLibrary,
		"note": "Status only. The pull runs as this account from a timer and the import runs as " +
			"the owning service's account; no tool here can trigger either, because the " +
			"fetcher deliberately cannot reach those databases. `last_pull` is the freshness " +
			"signal; `staged_now` is 0 in normal operation because the sync empties staging " +
			"after every run.",
	}, nil
}

// Handlers wires drive_status. Status takes no lock.
func Handlers(cfg *config.Config) map[string]server.ToolHandlerFunc {
	return map[string]server.ToolHandlerFunc{
		"drive_status": func(ctx context.Context, _ mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			out, err := Status(cfg, time.Now())
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			return mcpserver.ResultJSON(out)
		},
	}
}
